api/service.go

```diff
@@ -0,0 +1,466 @@
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/decred/dcrd/chaincfg/v3"
+	chainjson "github.com/decred/dcrd/rpc/jsonrpc/types/v3"
+	apiTypes "github.com/decred/dcrdata/v7/api/types"
+)
+
+type Backend int
+type Service struct {
+	client      *Client
+	chainParams *chaincfg.Params
+	backendUrl  map[string]map[Backend]string
+	urlMu       sync.Mutex
+}
+
+const (
+	Bittrex Backend = iota
+	Binance
+	BlockBook
+	DcrData
+	KuCoin
+)
+
+const (
+	testnetAddressIndetifier = "T"
+	mainnetAddressIdentifier = "D"
+	mainnetXpubIdentifier    = "d"
+	testnetXpubIdentifier    = "t"
+)
+
+var (
+	mainnetUrl = map[Backend]string{
+		Bittrex:   "https://api.bittrex.com/v3",
+		Binance:   "https://api.binance.com",
+		BlockBook: "https://blockbook.decred.org:9161/",
+		DcrData:   "https://mainnet.dcrdata.org/",
+		KuCoin:    "https://api.kucoin.com",
+	}
+
+	testnetUrl = map[Backend]string{
+		Binance:   "https://testnet.binance.vision",
+		BlockBook: "https://blockbook.decred.org:19161/",
+		DcrData:   "https://testnet.dcrdata.org/",
+		KuCoin:    "https://openapi-sandbox.kucoin.com",
+	}
+
+	supportedBackends = []string{"Bittrex", "Binance", "BlockBook", "DcrData", "KuCoin"}
+)
+
+func NewService(chainParams *chaincfg.Params) *Service {
+	conf := &ClientConf{
+		Debug: true,
+	}
+	client := NewClient(conf)
+	client.RequestFilter = func(info RequestInfo) (req *http.Request, err error) {
+		req, err = http.NewRequest(info.Method, info.Url, bytes.NewBuffer([]byte(info.Payload.(string))))
+		if err != nil {
+			log.Error(err)
+			return
+		}
+		if info.Method == "POST" || info.Method == "PUT" {
+			req.Header.Add("Content-Type", "application/json;charset=utf-8")
+		}
+		req.Header.Add("Accept", "application/json")
+
+		return
+	}
+
+	return &Service{
+		client:      client,
+		chainParams: chainParams,
+		backendUrl: map[string]map[Backend]string{
+			chaincfg.MainNetParams().Name:  mainnetUrl,
+			chaincfg.TestNet3Params().Name: testnetUrl,
+		},
+	}
+}
+
+// GetBestBlock returns the best block height as int32.
+func (s *Service) GetBestBlock() int32 {
+	s.setBackend(DcrData)
+	r, err := s.client.Do("GET", "api/block/best/height", "")
+	if err != nil {
+		log.Error(err)
+		return -1
+	}
+
+	h, err := strconv.ParseInt(string(r), 10, 32)
+	if err != nil {
+		log.Error(err)
+		return -1
+	}
+
+	return int32(h)
+}
+
+// GetBestBlockTimeStamp returns best block time, as unix timestamp.
+func (s *Service) GetBestBlockTimeStamp() int64 {
+	s.setBackend(DcrData)
+	r, err := s.client.Do("GET", "api/block/best?txtotals=false", "")
+	if err != nil {
+		log.Error(err)
+		return -1
+	}
+	var blockDataBasic *BlockDataBasic
+	err = json.Unmarshal(r, &blockDataBasic)
+	if err != nil {
+		log.Error(err)
+		return -1
+	}
```

Comment:
Its seems like this implementation is duplicated all through the methods.
I think creating a general function that is called by all of this duplicate implementations will reduce the code duplication. Consider having a such helper function:
``` golang
func (s *Service) queryExchange(exchange, method, url string, data interface) (error, interface) { 
          s.setBackend(exchange)
	  resp, err := s.client.Do(method, url, "")
	  if err != nil {
		return err, data
	}
	err = json.Unmarshal(resp, &data)
        // Its very important to first evaluate the data before returning the result.
        return err, data
	}
```
So now you can call `s.queryExchange(...)` wherever you see that whole is duplicated.

Reply:
This is the unique part that should be restricted in the specific function. Its actually possible to do it the way i suggested.

Reply:
Otherwise common implementation should be moved to a collective function